### PR TITLE
[stdlib] Rewrite integer parsing

### PR DESCRIPTION
**Pull request by PatrickPijnappel**
This is a rewrite of integer parsing for performance and code size. It address several major issues with the existing implementation (SR-7556), but also uses a new approach described below to achieve significant further gains.

Resolves SR-7556.

### [Updated] Latest Benchmark Results
By default, the benchmark suite only runs the tests marked in bold and skips the others.

| | | Decimal | Hex | Binary | Other |
|----------|:--------|:-------:|:-----:|:------:|:-----:|
|**Small** | **Int** | **1.5** | 1.7 | 2.5 |**1.0**|
| | **UInt** | 2.1 | 2.1 |**5.6** | 1.4 |
|**32-bit**| **Int** | 2.0 | 3.3 | 4.7 | 1.1 |
| | **UInt** | 1.9 | 3.2 | 8.0 | 1.0 |
|**64-bit**| **Int** | 3.5 | 6.3 | 7.2 | 1.4 |
| | **UInt** | **3.8** | **6.2** | 11.5 | 1.7 |

### Notes on Approach

The key idea here is using the bytes of a `UInt64` in a way similarly to SIMD. When doing ASCII operations the new UTF-8 backing of Strings makes them perfectly suited for this technique. Aside from integer parsing, for example also ASCII case manipulation sees many-X speed-ups when using this approach. Example (details below):

```swift
let str = 0x3132_3030_3633_3739 // ASCII "12006379"
let digits = str &- 0x3030_3030_3030_3030 // Subtract "0" from every lane
// … (underflow check omitted for clarity)
let c = (0x7f - 9) &* 0x0101_0101_0101_0101 // Constant to check value > 9
let isAnyAbove9 = (digits &+ c) & 0x8080_8080_8080_8080 != 0
```

In most cases, each byte is treated as a 7-bit value, with the remaining high bit per byte being considered a flag for tests and to capture under/overflow.

#### Operation Rules
- Bitwise operations (`|`, `&`, `^`, `~`) are clearly always valid, and shifts only if you know the lanes won't mix.
- When any of the flags bits are set, further operations except bitwise are not generally valid.
- Addition (when both values are 7-bit) is valid, with the flag bit indicating overflow.
- Subtraction (when both values are 7-bit) is valid, however upon underflow all more significant lanes will be garbled. The flag bit for the underflowing lane will be reliably set though.
- Comparisons (which set flags) can be made using `value &+ c` (`>`/`>=`) or `c &- value` (`<`/`<=`), by picking `c` such that for the `true` case the lane will overflow, i.e. be `>= 0x80`.
- The flags can be made into 7-bit masks using `flags = value & 0x8080…; mask = flags &- (flags &>> 7)`. This mask can then be used to branchlessly conditionalize further operations, e.g. `value &+= (…) & mask`.
- Multiplying a byte by `0x0101_0101_0101_0101` puts it in every lane.

### CODE REVIEW

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by stephentyrone** [comment]
Hi @theblixguy, this is exciting, and the benchmark numbers look promising!

I won't have time to do a detailed review until after Friday. @milseman, can you take a look at the stringy-aspects of this PR, as well?

File: stdlib/public/core/IntegerParsing.swift [hunk 1]
```diff
@@ -10,15 +10,404 @@
...
+  ///     Int("21-50")                      // Invalid format
+  ///     Int("ff6600")                     // Characters out of bounds
+  ///     Int("zzzzzzzzzzzzz", radix: 36)   // Out of range
+  ///
+  /// - Parameters:
+  ///   - text: The ASCII representation of a number in the radix passed as
+  ///     `radix`.
+  ///   - radix: The radix, or base, to use for converting `text` to an integer
+  ///     value. `radix` must be in the range `2...36`. The default is 10.
+  @inlinable @inline(__always)
```
Comment (tbkka) [inline stdlib/public/core/IntegerParsing.swift]: How do the perf numbers compare if you make this not inlineable at all?

Generally, I'd like to see us be much less aggressive about inlining; we're getting more feedback from folks that code size is a major concern and is in fact causing serious performance problems on RAM-constrained systems.  At a minimum, could we reduce this down to a top-level switch on radix and string implementation that selects among non-inlined functions? Then any given invocation should optimize down to just a single relevant function call.

File: stdlib/public/core/IntegerParsing.swift [hunk 2]
```diff
@@ -10,15 +10,404 @@
...
+  ///     Int("10000000000000000000000000") // Out of range
+  ///
+  /// - Parameter description: The ASCII representation of a number.
+  @inlinable
+  public init?(_ description: String) {
+    self.init(description, radix: 10)
+  }
+}
+
+@inlinable @inline(__always)
```
Comment (tbkka) [inline stdlib/public/core/IntegerParsing.swift]: See above.  It would be nice to avoid inlining functions at this level.

**Review by tbkka — COMMENTED** [review]
I couldn't find any tests for this in the current stdlib test suite other than some very basic sanity checks.  At a minimum, I would like to see:
* Test for each standard integer type (Int8, UInt8, ... , Int64, UInt64)
* Include positive and negative values
* Tests for min and max values
* Verify that min - 1 and max + 1 are correctly rejected
* Verify all special-cased radices (10, 16, 2) and at least one non-special-cased radix (3, 36)
* Test boundary values for this algorithm such as 9999 and 99999999
* Exercise short-string and non-short-string cases

File: stdlib/public/core/IntegerParsing.swift [hunk 3]
```diff
@@ -10,15 +10,404 @@
...
+      _fastPath(Result.isSigned && first == _ascii("-"))
+      || _fastPath(first == _ascii("+"))
+      else { return nil }
+    hasMinus = _fastPath(first == _ascii("-"))
+    word1 = (word1 ^ first) | 0x30 // Clear first char and a "0".
+  } else {
+    guard _fastPath(count > 0) else { return nil }
+  }
+  // Choose specialization based on radix (normally branch is optimized away).
+  let result_: UInt64?
```
Comment (stephentyrone) [inline stdlib/public/core/IntegerParsing.swift]: Given this, the implementation can't actually be generic over `Result: FixedWidthInteger`, because it doesn't work for a conforming type larger than `UInt64`, right?

**Review by stephentyrone — DISMISSED** [review]

File: stdlib/public/core/IntegerParsing.swift (hunk 3 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: It's fine for base 2/10/16, since the max number of characters in a small string is 15 and 16**15 < 2**64. The default case however could be as large as 36**15 so that one indeed shouldn't use that intermediate.

**Review by milseman — COMMENTED** [review]
I'll try to give a better code review later, but for now I just wanted to say I really appreciate you looking at this. The existing implementation was a lingering pain point and in general I like your approach.

File: stdlib/public/core/IntegerParsing.swift [hunk 4]
```diff
@@ -192,9 +81,315 @@ extension FixedWidthInteger {
...
+internal func _parseSmall<Result: FixedWidthInteger>(
+  _ rawGuts: _StringObject.RawBitPattern, radix: Int
+) -> Result? {
+  // We work with the bytes as they are arranged in the StringObject,
+  // i.e. the leading characters are in the least-significant bits.
+  // This is as opposed to SmallString order, which swaps them on big-endian
+  // platforms to maintain the same in-memory order.
+  var word1 = rawGuts.0
+  let word2 = rawGuts.1 & 0x00ff_ffff_ffff_ffff
+  let count = Int((rawGuts.1 &>> 56) & 0x0f)
```
Comment (milseman) [inline stdlib/public/core/IntegerParsing.swift]: You can get a `_SmallString` struct from the guts, which has some of these operations on it. In general, we'd want to avoid duplicating these magic flag values.

File: stdlib/public/core/IntegerParsing.swift (hunk 1 above)
Comment (milseman) [reply stdlib/public/core/IntegerParsing.swift]: We definitely want to be able to remove the generic type since that vast majority of call-sites will not be generic. The pattern we've often used is to have a small inlinable function that calls out to concrete implementation functions.

File: stdlib/public/core/IntegerParsing.swift (hunk 2 above)
Comment (milseman) [reply stdlib/public/core/IntegerParsing.swift]: This is definitely way too much inlining; but without inlining the generics don't go away. The fast-paths we care about are mostly around contiguous UTF-8 Strings/Substrings being parsed into results that fit in 64-bits. The special path for small strings might be warranted. However, we should be able to strip all generics fairly early on for fast paths and invoke non-inlined code.

**Comment by PatrickPijnappel** [comment]
Yes definitely this inlining is not shippable as-is. Best approach I'm thinking now would be have a non-generic `parseSmallDec/Hex/Bin -> UInt64` + non-generic `parseFromContiguousUTF8 -> UInt64` + generic `parseFromUTF8Iterator`.

However it'd be really valuable if we could expand the specialized dec/hex/bin methods to any contiguous UTF8 (given <= UInt64). Especially since the small case atm doesn't cover substrings, which would be very common when parsing from text files.

The issue is efficiently loading the chunks. I've been thinking in most cases we can guarantee it's OK to load aligned UInt64 chunks from `utf8Ptr & ~7` through `(utf8Ptr + count - 1) & ~7` right – i.e. all words in which there's character we need to read. This should be the case for all small strings and if I understand correctly for all native-backed strings (note: not a 100% about the 32-bit case). Bridged strings don't give any guarantees AFAIK, so those would be out. @milseman Thoughts?

**Comment by benrimmington** [comment]
> I couldn't find any tests for this in the current stdlib test suite other than some very basic sanity checks. At a minimum, I would like to see:
> 
> * Test for each standard integer type (Int8, UInt8, ... , Int64, UInt64)
> * Include positive and negative values
> * Tests for min and max values
> * Verify that min - 1 and max + 1 are correctly rejected
> * Verify all special-cased radices (10, 16, 2) and at least one non-special-cased radix (3, 36)
> * Test boundary values for this algorithm such as 9999 and 99999999
> * Exercise short-string and non-short-string cases

@tbkka I think most of these can be found in [test/stdlib/NumericParsing.swift.gyb](https://github.com/apple/swift/blob/f1fa041bd4476924032984c8b1ea997671d4123a/test/stdlib/NumericParsing.swift.gyb#L54-L132)

**Comment by PatrickPijnappel** [comment]
@benrimmington @tbkka Yeah there a few there already, however I'm currently writing tests to cover more cases, including ones mentioned that are missing and several more. This rewrite introduces some new code paths and boundary values that should be explicitly covered, and the original tests were fairly limited generally.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

File: stdlib/public/core/IntegerParsing.swift [hunk 5]
```diff
@@ -192,9 +95,362 @@ extension FixedWidthInteger {
...
+) -> UInt64? {
+  return _parseIntegerSpecialized(from: utf8, allowNegative: allowNegative,
+                                  max: max, using: { utf8 in
+    var iter = utf8.makeIterator()
+    let first = iter.next()._unsafelyUnwrappedUnchecked
+    return _parseFromUTF8Unsigned(first: first, rest: &iter, radix: radix)
+  })
+}
+
+@inline(__always)
```
Comment (tbkka) [inline stdlib/public/core/IntegerParsing.swift]: I'd like to see the benchmark results if this function and the following `_parseUnsignedBaseXX` functions were all `@usableFromInline` instead of `@inline(__always)`.  Could you try that experiment?  I suspect that you would still see a healthy speedup over the old version with dramatically smaller code size in the clients.

Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: Note that these functions are not ABI-public, they are never inlined in the client only in the standard library itself. They are just a convenience to generate the specialized @usableFromInline methods above.

Comment (tbkka) [reply stdlib/public/core/IntegerParsing.swift]: Ah.  Of course, that makes perfect sense.  Might be worth checking whether these annotations are really needed; I'm not an expert in compiler optimizations, but I have the impression that it does a pretty good job of inlining within a module.

File: stdlib/public/core/IntegerParsing.swift [hunk 6]
```diff
@@ -146,31 +43,37 @@ extension FixedWidthInteger {
...
-      }) else {
-        return nil
-      }
-      self = ret
-      return
+    precondition(2...36 ~= radix)
+    let wholeGuts = text._wholeGuts
+    // The specialized paths require that all of the contiguous bytes can
+    // be read using UInt64 loads from (address & ~7).
+    if wholeGuts._object.isPreferredRepresentation, Self.bitWidth <= 64 {
```
Comment (PatrickPijnappel) [inline stdlib/public/core/IntegerParsing.swift]: @milseman Would this be the correct check to guarantee `_loadUnalignedChunk` is valid? It's clear valid for small strings, and AFAICT also for natively stored large strings (both on 32 & 64-bits), not so sure about literals.

Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: Or, if we can perhaps even guarantee this for shared and/or bridged UTF-8?

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke-test

File: stdlib/public/core/IntegerParsing.swift (hunk 6 above)
Comment (milseman) [reply stdlib/public/core/IntegerParsing.swift]: I don't think we can make this assumption. At the very least, a `CFString` may come in unaligned. CC @stephentyrone . We want better utilities for dealing with larger-strides alignment in general.

File: stdlib/public/core/IntegerParsing.swift [hunk 7]
```diff
@@ -146,31 +43,37 @@ extension FixedWidthInteger {
...
+      let r64_ = wholeGuts.withFastUTF8({ utf8 -> UInt64? in
```
Comment (milseman) [inline stdlib/public/core/IntegerParsing.swift]: I wonder if this would be significantly simpler to just force-contiguous the storage up front. This would incur a copy for non-small lazily bridged NSStrings. But longer term, I think this is a worthwhile tradeoff.

```swift
var copy = text
copy.withUTF8 { ... }
```

which will also handle the slicing for you (easy to forget).

File: stdlib/public/core/IntegerParsing.swift (hunk 6 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: Note that the requirement isn't alignment exactly, but rather that the memory at the address rounded down to a 8-byte multiple (even on 32-bit systems) is valid to read even if it's garbage.

Per string backing:
- Small: should be fine
- Large (native, mortal): should be fine on both 32 & 64 bit AFAICT
- Large, immortal: not sure about this
- Shared & bridged: Probably not?
- Foreign: no

File: stdlib/public/core/IntegerParsing.swift (hunk 7 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: That does sounds like it might be the best approach. Would this give us the desired guarantee? Being, that the following is allowable for every code unit:
```swift
let chunk = UnsafeRawPointer(bitPattern: p & ~0x07)!.load(as: UInt64.self)
```

File: stdlib/public/core/IntegerParsing.swift (hunk 6 above)
Comment (stephentyrone) [reply stdlib/public/core/IntegerParsing.swift]: This assumption does not hold. Swift hasn't yet defined a formal memory model that addresses this detail, but absent a guarantee, it is UB to read outside the bounds of an object, *even if that's an aligned read, even if it would be allowed by the hardware*. If we want to take advantage of hardware support for doing this, it has to be done in assembly at present (C and C++ make this explicitly UB, Swift may or may not in the future, so we can't depend on the semantics at this point).

Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: For objects in general yes, but for some specific string backing it should be valid. In particular (correct me if I'm wrong):
- Small: UTF-8 pointer points to 2 UInt64's on the stack (on both 32 & 64-bit arch), so the assumption holds
- Large (native, mortal): UTF-8 pointer is tail allocated memory with header of 20/32 bytes (32/64 bits) and for <128 bytes the capacity is bucketed to 16 byte multiples. Above that it should have breadcrumbs, which is word size. That only leaves a question for 32-bit platforms?
Not so sure about the memory layout of literals.

Comment (tbkka) [reply stdlib/public/core/IntegerParsing.swift]: The most robust approach here would be to take the fast(est) path if the pointer is suitably aligned, else fall back to a byte-by-byte approach.  In particular, that will be safe even if the string internals change in the future.

Comment (stephentyrone) [reply stdlib/public/core/IntegerParsing.swift]: If it's suitably aligned, then it's valid to read from, but then there's no need to round down, because it's suitably aligned.

Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: @stephentyrone does the above reasoning for small & native string not hold then? I'm trying to find a subset for which we can have this guarantee because the performance difference is rather large, and could be generalized to other string operations such as upper/lowercasing.

@tbkka Note there's also the question of reading beyond the end. And the perf difference between byte-by-byte approach and chunk-by-chunk is quite large so if at all possible it'd be worth it.

Comment (stephentyrone) [reply stdlib/public/core/IntegerParsing.swift]: @PatrickPijnappel There's no need to go byte-by-byte or read beyond the end; both misalignment and trailing bytes can be handled without significant performance penalty and without reading any memory outside the string bounds:

- If small, just load two words and check them, masking out unused bytes with a shift.
- If 8 bytes or less, and not small, go byte by byte (this case is quite rare, perf doesn't matter much)
- Otherwise:
  - Load the first 8 bytes of the string, regardless of alignment. This is not quite expressible in Swift today, so we'll have to define an unaligned load function in the stubs header that does it (this is trivial, happily).
  - Advance to the next aligned location in the string, `start+8 & -8`. Subtract the distance from this pointer and `start` from length.
  - As long as length > 8, load aligned words.
  - Do a final unaligned load from `end-8`.

This does two unaligned loads, but all other accesses are aligned, and no byte-by-byte access, and never reads outside the valid string bounds. On modern x86 and arm implementations the two unaligned accesses are fast (and the first one will actually be aligned in practice almost every time).

Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: That makes a lot of sense, I'll implement that this weekend! Can you give some pointers to adding the unaligned load to the stub header?

File: stdlib/public/core/IntegerParsing.swift (hunk 7 above)
Comment (milseman) [reply stdlib/public/core/IntegerParsing.swift]: Sorry, is that addressed to me? If so, could you rephrase it? (I apologize for losing track of this PR last week)

File: stdlib/public/core/IntegerParsing.swift (hunk 6 above)
Comment (tbkka) [reply stdlib/public/core/IntegerParsing.swift]: @PatrickPijnappel -- I think @stephentyrone is referring to this header file: `stdlib/public/SwiftShims/RuntimeShims.h` and the associated implementations which seem to be mostly in `stdlib/public/stubs/Stubs.cpp`.  These files (and the other files in those directories) provide various small C/C++ support functions to the standard library, many of which serve to hide processor- or operating-system-specific details.

File: stdlib/public/core/IntegerParsing.swift (hunk 7 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: @milseman No worries, it was but resolved it with stephen canon above (since this guarantee doesn't always hold).

File: stdlib/public/core/IntegerParsing.swift (hunk 6 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: Do you know whether this should be an inline function or whether that doesn't matter for the compiler? The stubs seem to be defined a couple of different places, e.g. RuntimeStubs.h, RuntimeShims.h and Runtime.swift (in stdlib), which is the proper place?

Implementation-wise do we just need
```cpp
uint64_t loadUInt64Unaligned(void *p) {
  return *(uint64_t *)p;
}
```

Comment (stephentyrone) [reply stdlib/public/core/IntegerParsing.swift]: No, that generates an aligned load; by casting the pointer to `uint64_t`, you are telling the compiler that it is suitably aligned for an 8-byte integer.

You want:
```c
static inline SWIFT_ALWAYS_INLINE
__swift_uint64_t loadUInt64Unaligned(char *p) {
  __swift_uint64_t result;
  memcpy(&result, p, sizeof result);
  return result;
}
```

Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift]: @stephentyrone The `memcpy` appears to hit some bug in SIL deserialization: https://gist.github.com/PatrickPijnappel/f133f7296eb05b966a7bc9486ded35b7

Implemented it using the alignment attribute for now

**Comment by PatrickPijnappel** [comment]
@milseman @stephentyrone This should now be all done, except for an issue with the benchmarks: there's retains/releases happening around the integer parsing call taking up to 70% of the call ([assembly](https://gist.github.com/PatrickPijnappel/1b55770147ae952e5d5b374ee9ca1ea0)), even though the strings are loaded from a constant global array. This shouldn't be necessary as far as I can tell but I can't seem to figure out why the retains are not optimized away, any ideas?

```swift
for string in uint64DecimalStrings {
  result &+= UInt64(string, radix: 10)!
}
```

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

File: stdlib/public/core/IntegerParsing.swift (hunk 6 above)
Comment (stephentyrone) [reply stdlib/public/core/IntegerParsing.swift]: Ah, right, forgot that you can't use `stdlib` functions in the shims. If you use `__builtin_memcpy` instead it will work.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@milseman Would it be valid to mark the underlying non-inlineable implementations as `@_effects(readonly)`?
```swift
@usableFromInline @_effects(readonly)
internal func _parseIntegerBase10(
  from utf8: UnsafeBufferPointer<UInt8>, allowNegative: Bool, max: UInt64
) -> UInt64? {
  // …
}
```
As far as I can tell from the [docs](https://github.com/apple/swift/blob/master/docs/HighLevelSILOptimizations.rst#_effects-attribute) that sounds like it should be valid, and gets rid of the unnecessary ref counting.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@eeckstein Any idea why these string equality benchmarks are consistently down? I can't find any touchpoint between the two

**Comment by PatrickPijnappel** [comment]
@milseman @stephentyrone This is now ready for review! Note that I added the `@_effects(readonly)` that'd I'd like to double check is valid.

Full updated benchmark numbers (speed-up multiples):
| | | Decimal | Hex | Binary | Other |
|----------|:--------|:-------:|:-----:|:------:|:-----:|
|**Small** | **Int** | **1.5** | 1.7 | 2.5 |**1.0**|
| | **UInt** | 2.1 | 2.1 |**5.6** | 1.4 |
|**32-bit**| **Int** | 2.0 | 3.3 | 4.7 | 1.1 |
| | **UInt** | 1.9 | 3.2 | 8.0 | 1.0 |
|**64-bit**| **Int** | 3.5 | 6.3 | 7.2 | 1.4 |
| | **UInt** | **3.8** | **6.2** | 11.5 | 1.7 |

File: stdlib/public/core/IntegerParsing.swift [hunk 8]
```diff
@@ -146,31 +45,44 @@ extension FixedWidthInteger {
...
+        &- text.startIndex._encodedOffset >= 9
+    }
+    if _fastPath(canChunkLoad), _fastPath(wholeGuts.isFastUTF8),
+        Self.bitWidth <= 64 {
+      let r64_ = wholeGuts.withFastUTF8({ utf8 -> UInt64? in
+        let utf8 = (S.self == String.self) ? utf8 :
+          UnsafeBufferPointer(rebasing: utf8[text._offsetRange])
+        let isSigned = Self.isSigned
+        let max = UInt64(Self.max)
+        switch radix {
```
Comment (stephentyrone) [inline stdlib/public/core/IntegerParsing.swift:67]: Side note for anyone following along; boy would it would be swell if we could precisely express the semantics that this switch should always be inlined when it's going to get eliminated, and never inlined when it won't.

**Comment by tbkka** [comment]
I would love to reduce code size in the `-O` case a bit further.  If we could do that without giving up too much perf, I think it would be worthwhile.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

File: stdlib/public/core/IntegerParsing.swift [hunk 9]
```diff
@@ -146,31 +45,44 @@ extension FixedWidthInteger {
...
+        default: return _parseInteger(from: utf8, radix: radix,
+                                      allowNegative: isSigned, max: max)
+        }
+      })
+      guard _fastPath(r64_ != nil), let r64 = r64_ else { return nil }
+      self = Self(truncatingIfNeeded: r64)
+    } else {
+      var iter = text.utf8.makeIterator()
+      let result_: Self? = _parseInteger(from: &iter, radix: radix)
+      guard _fastPath(result_ != nil), let result = result_ else { return nil }
```
Comment (milseman) [inline stdlib/public/core/IntegerParsing.swift:83]: Are you seeing any codegen differences from some of the more liberal uses of `_fastPath`?

**Review by milseman — COMMENTED** [review]
Hoping to simplify and remove some of the code code paths if possible.

**Comment by PatrickPijnappel** [comment]
> I would love to reduce code size in the `-O` case a bit further. If we could do that without giving up too much perf, I think it would be worthwhile.

@tbkka I'm not sure we can, significantly. The main initializer is the only one that is inlinable, and it does seem to require `@inline(__always)` to reliable eliminate the switch. Note that the most common case in practice, the initializer without radix, is not `@inline(__always)` resulting in significant reductions, i.e. see the reduction for `StrToInt.o`.

File: stdlib/public/core/IntegerParsing.swift (hunk 9 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift:83]: I primarily added them for branch hints, do you have any concern regarding adding these?

File: stdlib/public/core/IntegerParsing.swift [hunk 10]
```diff
@@ -146,31 +45,44 @@ extension FixedWidthInteger {
...
+    // Check we can load the UTF-8 in UInt64 chunks.
+    let wholeGuts = text._wholeGuts
+    let canChunkLoad: Bool
+    if S.self == String.self {
+      canChunkLoad = wholeGuts.isSmall || wholeGuts.count >= 9 // +1 for sign.
+    } else {
+      canChunkLoad = (wholeGuts.isSmall ? 16 : wholeGuts.count)
+        &- text.startIndex._encodedOffset >= 9
+    }
+    if _fastPath(canChunkLoad), _fastPath(wholeGuts.isFastUTF8),
```
[the remark this answers is not in the document]
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift:60]: I think there's definitely a good case for removing the specialization for the default case here, as that one is probably not worth it – even though it would mean uncommon radix performance would be regressing.

We need the separate generic implementation not just because of cases lacking fast UTF-8 but also for those that don't fit into a UInt64.

In terms of forcing UTF-8, that might be worth it, and maybe even copying in the less-than 9 char non-smol case. So what about this line-up:
- Base 10/2/16, type fits UInt64: force UTF-8 (and smol if count < 9) and use a radix specialized path
- Uncommon radix, or type > UInt64: generic slow path

File: stdlib/public/core/IntegerParsing.swift (hunk 9 above)
Comment (PatrickPijnappel) [reply stdlib/public/core/IntegerParsing.swift:83]: Oh just noticed this particular one is completely redundant. Must've been left over from some refactoring.

**Comment by PatrickPijnappel** [comment]
@stephentyrone Side note, will the unsigned implementations (e.g. `_parseUnsignedBase10`) still exist separately in the stdlib, code size wise? Their only purpose is being inlined in the ABI-public methods above

If these aren't stripped automatically I can make them inlined closures instead.

**Comment by eeckstein** [comment]
@PatrickPijnappel It could just be a code alignment issue, which unfortunately sometimes gives "random" results.
To verify, you can compile the benchmarks locally and see if the generated assembly code really differs.

File: stdlib/public/core/IntegerParsing.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (milseman) [reply stdlib/public/core/IntegerParsing.swift:60]: Right, a generic unspecialized fall-back path would be slow. I'm not too concerned whether that path forces contiguous UTF-8, and it seems that instead just yielding an iterator is sufficient. The specialized fast paths might as well force contiguous UTF-8 IMO.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
The benchmarks seem to have been accidentally deleted in another PR, re-adding them here: #31326.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@milseman I did a bit of a reorganization to simplify by forcing UTF-8. However, two issues:
- The `withUTF8`, at least in the way it's currently implemented, is causing retain/release that's traffic killing performance. Any ideas?
- The uncommon radix performance is more than an order of magnitude worse than master.

File: stdlib/public/core/IntegerParsing.swift [hunk 11]
```diff
@@ -146,31 +45,54 @@ extension FixedWidthInteger {
...
+      canChunkLoad = (wholeGuts.isSmall ? 16 : wholeGuts.count)
+        &- text.startIndex._encodedOffset >= 9
+    }
+    // Wrapper to handle the fact that withUF8 is defined on both
+    // String and Substring, but not on StringProtocol.
+    let withUTF8: ((UnsafeBufferPointer<UInt8>) -> Self?) -> Self? = { f in
+      if S.self == String.self {
+        var text = text as! String
+        return text.withUTF8(f)
+      } else { // StringProtocol requires no additional types conform to it.
```
Comment (karwa) [inline stdlib/public/core/IntegerParsing.swift:67]: This may not be correct. I think we may technically be leaving the door open to having additional types conform to StringProtocol. In any case, it is worth considering if we want to expose this in an `inlinable` function.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by PatrickPijnappel** [comment]
> Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically.
> 
> * How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
> * More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

Updated. I unfortunately haven't had the time to get back to finishing this up, but hope to have some time soon.
